### PR TITLE
chore: Refactor to allow sharing data source status and IDataSource.

### PR DESCRIPTION
**Pull request by kinyoklion**
This should be backwards compatible. The main validation of this is that the bindings are not updated. Many times along the journey there were things that would have required binding changes.

### CODE REVIEW

File: libs/client-sdk/tests/data_source_status_test.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include <gtest/gtest.h>
+
+#include <launchdarkly/client_side/data_source_status.hpp>
+
+using launchdarkly::client_side::data_sources::DataSourceStatus;
+
+TEST(DataSourceStatusTest, OstreamBasicStatus) {
```
Comment (kinyoklion) [inline libs/client-sdk/tests/data_source_status_test.cpp:7]: Basically everything else is tested through the data source status manager.

File: libs/common/include/launchdarkly/data_sources/data_source_status_base.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+#include <launchdarkly/data_sources/data_source_status_error_info.hpp>
+#include <launchdarkly/data_sources/data_source_status_error_kind.hpp>
+
+// Common is included in the namespace to disambiguate from client/server
+// for backward compatibility.
+namespace launchdarkly::common::data_sources {
+
+template <typename TDataSourceState>
+class DataSourceStatusBase {
```
Comment (kinyoklion) [inline libs/common/include/launchdarkly/data_sources/data_source_status_base.hpp:12]: Says Base, but it isn't a base now. But it needs a name different than DataSourceStatus, as to not cause an aneurism.

File: libs/client-sdk/tests/data_source_status_test.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    auto time =
+        std::chrono::system_clock::time_point{std::chrono::milliseconds{0}};
+
+    DataSourceStatus status(DataSourceStatus::DataSourceState::kInitializing,
+                            time, std::nullopt);
+
+    std::stringstream ss;
+    ss << status;
+    ss.flush();
```
Comment (cwaldren-ld) [inline libs/client-sdk/tests/data_source_status_test.cpp:16]: I believe `flush` is a no-op for stringstreams

**Review by cwaldren-ld — APPROVED** [review]
Aside from the lints, LGTM!

File: libs/client-sdk/include/launchdarkly/client_side/data_source_status.hpp [hunk 4]
```diff
@@ -5,202 +5,73 @@
...
-
-    DataSourceStatus(DataSourceStatus const& status);
-
-   private:
-    DataSourceState state_;
-    DateTime state_since_;
-    std::optional<ErrorInfo> last_error_;
+    // BackgroundDisabled,
+    // TODO: A plugin of sorts would likely be required for some
+    // functionality like this. kNetworkUnavailable,
```
[the remark this answers is not in the document]
Comment (kinyoklion) [reply libs/client-sdk/include/launchdarkly/client_side/data_source_status.hpp:69]: kNetworkUnavailable is a commented out enum. I am not sure if maybe formatting wrapped it into the comment.

**Pull request merged**